### PR TITLE
feat(arch): add stm32 support

### LINKED ISSUE #168: STM32 Support
There are so many boards to support from STMicro, but one that sticks out is STM32WB55XX with its radio ?

### PR DESCRIPTION
**Pull request by kaspar030**
This PR is my WIP branch for wiring up stm32 support.

I initialy started supporting the Nintendo Game and Watch handheld. That's actually a pretty useless target without quite a bit more work (like, proper display support). I got it to print and stopped there. I'll leave it in.

More interesting is the support for the ST Nucleo H755ZI-Q which I had on my desk.
That currently freezes in `embassy_stm32::init()`, I'm investigating.

**update** I've added basic nucleo-f401re and nucleo-wb55 support. Need to find a way to configure the used software interrupt.

** update **

- [x] find way to configure the used software interrupt => #285 
- [ ] find place for per-MCU `Config`

Closes #168.

### CODE REVIEW

File: src/riot-rs-embassy/src/arch/dummy.rs [hunk 1]
```diff
@@ -195,3 +195,6 @@ pub mod usb {
         }
     }
 }
+
+// warn if this is used
+fn warning_using_dumma_arch() {}
```
Comment (maribu) [inline src/riot-rs-embassy/src/arch/dummy.rs:200]:
```suggestion
fn warning_using_dummy_arch() {}
```

?

File: laze-project.yml [hunk 2]
```diff
@@ -294,6 +294,66 @@ contexts:
...
+    parent: stm32
+    selects:
+      - thumbv7em-none-eabi # actually eabihf, but riot-rs doesn't support hard float yet
+    env:
+      PROBE_RS_CHIP: STM32WB55RGVx
+      PROBE_RS_PROTOCOL: swd
+      RUSTFLAGS:
+        - --cfg context=\"stm32wb55rgvx\"
+      CARGO_ENV:
+        - CONFIG_SWI=LPUART1
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply laze-project.yml:334]: I've disabled automatic use of the interrupt executor for stm32.

File: src/riot-rs-boards/Cargo.toml [hunk 3]
```diff
@@ -22,11 +22,14 @@ nrf52840-mdk = { optional = true, path = "nrf52840-mdk" }
...
+st-nucleo-f401re = { optional = true, path = "st-nucleo-f401re" }
+st-nucleo-wb55 = { optional = true, path = "st-nucleo-wb55" }
 rpi-pico = { optional = true, path = "rpi-pico" }
 particle-xenon = { optional = true, path = "particle-xenon" }
+st-nucleo-h755zi-q = { optional = true, path = "st-nucleo-h755zi-q" }
 
 [features]
 # Allows to have no boards selected, useful to run platform-independent tooling
 no-boards = []
+# both "rpi-pico" and "rpi-pico-w" are valid, so no "dep:rpi-pico"
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-boards/Cargo.toml:34]: 🫣

File: src/riot-rs-embassy/src/arch/stm32/mod.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        config.rcc.apb3_pre = APBPrescaler::DIV2; // 100 Mhz
+        config.rcc.apb4_pre = APBPrescaler::DIV2; // 100 Mhz
+        config.rcc.voltage_scale = VoltageScale::Scale1;
+        // Set SMPS power config otherwise MCU will not powered after next power-off
+        config.rcc.supply_config = SupplyConfig::DirectSMPS;
+        config.rcc.mux.usbsel = mux::Usbsel::HSI48;
+    }
+
+    // mark used
+    let _ = config;
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/arch/stm32/mod.rs:87]: yes, there's no special case for the nucleo-f401re, without this there's a warning

File: src/riot-rs-embassy/src/arch/stm32/usb_synopsis_otg.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    let ep_out_buffer = crate::make_static!([0u8; 256]);
+    let mut config = embassy_stm32::usb::Config::default();
+
+    // Enable vbus_detection
+    // Note: some boards don't have this wired up and might not require it,
+    // as they are powered through usb!
+    // If you hang on boot, try setting this to "false"!
+    // See https://embassy.dev/book/dev/faq.html#_the_usb_examples_are_not_working_on_my_board_is_there_anything_else_i_need_to_configure
+    // for more information
+    config.vbus_detection = true;
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/arch/stm32/usb_synopsis_otg.rs:25]: yes. let's take this as is for now until we have a board that won't work?

File: src/riot-rs-embassy/src/arch/stm32/mod.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+#[cfg(feature = "executor-interrupt")]
+pub(crate) use embassy_executor::InterruptExecutor as Executor;
+
+use embassy_stm32::Config;
+
+#[cfg(feature = "usb")]
+cfg_if::cfg_if! {
+    if #[cfg(feature = "stm32_usb")] {
+        #[path = "usb.rs"]
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/arch/stm32/mod.rs:11]: it is redundant, I left it so it is consistent with the other case.

**Comment by kaspar030** [comment]
> Haven't tested it in hardware yet, what would be an appropriate example to try out?

well, any of the three wired-up nucleos should work, only usb-ethernet on the stm32h7 is currently not working correctly.

**Comment by kaspar030** [comment]
@ROMemories I think I got them all

**Pull request merged**